pkg/driverbuilder/builder/ubuntu.go

```diff
@@ -296,69 +205,85 @@ func fetchUbuntuAWSKernelURLS(baseURL string, kr kernelrelease.KernelRelease, ke
 			kr.Architecture.String(),
 		),
 	}
-}
 
-func parseUbuntuAWSKernelURLS(baseURL string, kr kernelrelease.KernelRelease, kernelVersion uint16) ([]string, error) {
-	resp, err := http.Get(baseURL)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
-	}
-	firstExtra := extractExtraNumber(kr.Extraversion)
-	rmatch := `href="(linux(?:-aws-%d.%d)?-headers-%s-%s(?:-aws)?_%s-%s\.%d.*(?:%s|all)\.deb)"`
-	fullRegex := fmt.Sprintf(rmatch, kr.Version, kr.PatchLevel,
-		kr.Fullversion, firstExtra, kr.Fullversion,
-		firstExtra, kernelVersion, kr.Architecture.String())
-	pattern := regexp.MustCompile(fullRegex)
-	matches := pattern.FindAllStringSubmatch(string(body), 2)
-	if len(matches) != 2 {
-		return nil, fmt.Errorf("kernel headers and kernel headers common not found")
+	// combine it all together now
+	packageFullURLs := []string{}
+	for _, url := range possibleFullURLs {
+		for _, packageName := range packageNamePatterns {
+			packageFullURLs = append(
+				packageFullURLs,
+				fmt.Sprintf("%s/%s", url, packageName),
+			)
+		}
 	}
 
-	foundURLs := []string{fmt.Sprintf("%s/%s", baseURL, matches[0][1])}
-	foundURLs = append(foundURLs, fmt.Sprintf("%s/%s", baseURL, matches[1][1]))
-	return foundURLs, nil
+	// return out the deduplicated url list
+	return deduplicateURLs(packageFullURLs), nil
 }
 
-func extractExtraNumber(extraversion string) string {
-	firstExtraSplit := strings.Split(extraversion, "-")
-	if len(firstExtraSplit) > 0 {
-		return firstExtraSplit[0]
+// deduplicate the array of URLs to ensure we are
+// only get unique resolving URLs for packages
+func deduplicateURLs(urls []string) []string {
+	keys := make(map[string]bool)
+	dedupURLs := []string{}
+
+	// loop over the URL list
+	// set a flag for new URLs in list, add to dedup list
+	// do nothing if URL is duplicate
+	for _, url := range urls {
+		if _, value := keys[url]; !value {
+			keys[url] = true
+			dedupURLs = append(dedupURLs, url)
+		}
 	}
-	return ""
+	return dedupURLs
 }
 
-type ubuntuTemplateData struct {
-	DriverBuildDir       string
-	ModuleDownloadURL    string
-	KernelDownloadURLS   []string
-	KernelLocalVersion   string
-	KernelHeadersPattern string
-	ModuleDriverName     string
-	ModuleFullPath       string
-	BuildProbe           bool
-	BuildModule          bool
-	GCCVersion           string
+// parse the extraversion from the kernelrelease to retrieve the extraNumber and flavor
+// assume the flavor is "generic" if unable to parse the flavor
+// Example: Input -> "188-generic", Output -> "188", "generic"
+// NOTE: make sure the kernelrelease passed in appears *exactly* as `uname -r` output
+func parseUbuntuExtraVersion(extraversion string) (string, string) {
```

Comment:
I think we should probably add a unit test for this, since these kernel versions will get pretty complex, as we've already seen.

Also, thanks for doing the leg work to get this function written! It was quite an adventure finding all the different flavors (I'm looking at you, `ubuntu-oracle`).

Reply:
Will do, on it boss 👍

Reply:
I wrote a starting point test file for the Ubuntu builder, was able to to achieve ~67% coverage. The tests aren't perfect, but they're a start.